File: Server/src/main/content/global/skill/gathering/hunter/TrapHook.java
package content.global.skill.gathering.hunter;

import core.game.world.GameWorld;
import core.game.world.map.Location;
import core.tools.RandomFunction;

/**
 * The Trap hook.
 */
public class TrapHook {

    private final TrapWrapper wrapper;

    private final Location[] locations;

    /**
     * Instantiates a new Trap hook.
     *
     * @param wrapper   the wrapper
     * @param locations the locations
     */
    public TrapHook(TrapWrapper wrapper, Location[] locations) {
        this.wrapper = wrapper;
        this.locations = locations;
    }

    /**
     * Gets chance location.
     *
     * @return the chance location
     */
    public Location getChanceLocation() {
        final double chance = wrapper.getChanceRate();
        final int roll = RandomFunction.random(99);
        final double successChance = (GameWorld.getSettings().isDevMode() ? 100 : 55.0) + chance;
        if (successChance > roll) {
            return RandomFunction.getRandomElement(locations);
        }
        return null;
    }

    /**
     * Is hooked boolean.
     *
     * @param location the location
     * @return the boolean
     */
    public boolean isHooked(Location location) {
        for (Location l : locations) {
            if (l.equals(location)) {
                return true;
            }
        }
        return false;
    }

    /**
     * Gets wrapper.
     *
     * @return the wrapper
     */
    public TrapWrapper getWrapper() {
        return wrapper;
    }

    /**
     * Get locations location [ ].
     *
     * @return the location [ ]
     */
    public Location[] getLocations() {
        return locations;
    }

}

File: Server/src/main/content/global/skill/gathering/hunter/TrapHook.kt
package content.global.skill.gathering.hunter

import core.game.world.GameWorld.settings
import core.game.world.map.Location
import core.tools.RandomFunction

/**
 * Trap hook
 *
 * @property wrapper This property holds a reference to the TrapWrapper instance.
 * @property locations This property contains an array of Location instances.
 * @constructor Trap hook Initializes a new instance of TrapHook with the specified wrapper and locations.
 */
class TrapHook(val wrapper: TrapWrapper, val locations: Array<Location>) {

    val chanceLocation: Location?
        get() {
            val chance = wrapper.chanceRate
            val roll = RandomFunction.random(99)
            val successChance = (if (settings!!.isDevMode) 100.0 else 55.0) + chance
            if (successChance > roll) {
                return RandomFunction.getRandomElement(locations)
            }
            return null
        }

    /**
     * Is hooked
     *
     * @param location
     * @return
     */
    fun isHooked(location: Location): Boolean {
        for (l in locations) {
            if (l == location) {
                return true
            }
        }
        return false
    }
}
